### PR TITLE
Box Developer Edition

### PR DESCRIPTION
**Pull request by bgoldman**
Implementing https://github.com/box/box-java-sdk/issues/149

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @bgoldman, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

**Comment by bgoldman** [comment]
CLA signed

**Comment by boxcla** [comment]
Verified that @bgoldman has just signed the CLA. Thanks, and we look forward to your contribution.

**Comment by bgoldman** [comment]
Not sure what's up with the Travis errors. If you have any idea, let me know.

File: build.gradle [hunk 1]
```diff
@@ -8,7 +8,7 @@ sourceCompatibility = 1.6
 
 group = 'com.box'
 archivesBaseName = 'box-java-sdk'
-version = '1.1.0'
+version = '1.1.1'
```
Comment (gcurtis) [inline build.gradle]: Let's not bump the version yet, we'll do that in another commit when we release.

File: src/main/java/com/box/sdk/BoxAPIConnection.java [hunk 2]
```diff
@@ -123,6 +140,46 @@ public static BoxAPIConnection restore(String clientID, String clientSecret, Str
...
+     * Authenticates the API connection for Box Developer Edition.
+     * @param entityId           An enterprise ID or a user ID.
+     * @param entityType         "enterprise" or "user", corresponding to entityId.
+     * @param privateKey         the private key corresponding to the public key configured with Box Developer Edition.
+     * @param privateKeyPassword the password for the private key.
+     * @throws JoseException
+     * @throws IOException
+     */
+    public void authenticate(String entityId, String entityType, String privateKey, String privateKeyPassword)
+        throws JoseException, IOException {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: I'd prefer that these exceptions be wrapped with a `BoxAPIException` for consistency. This way we don't have to expose classes from dependencies through the SDK.

File: src/main/java/com/box/sdk/BoxAPIConnection.java [hunk 3]
```diff
@@ -383,12 +440,54 @@ public void setProxyPassword(String proxyPassword) {
...
+     * @param clientId           the client ID to use when refreshing the access token.
+     * @param clientSecret       the client secret to use when refreshing the access token.
+     * @param privateKey         the private key corresponding to the public key configured with Box Developer Edition.
+     * @param privateKeyPassword the password for the private key.
+     * @return a new instance of BoxAPIConnection.
+     * @throws JoseException
+     * @throws IOException
+     */
+    public static BoxAPIConnection getAppAuthConnection(String enterpriseId, String clientId, String clientSecret,
+        String privateKey, String privateKeyPassword) throws JoseException, IOException {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: Same as above. Update exceptions.

File: src/main/java/com/box/sdk/BoxAPIConnection.java [hunk 4]
```diff
@@ -383,12 +440,54 @@ public void setProxyPassword(String proxyPassword) {
...
+     * @param clientId           the client ID to use when refreshing the access token.
+     * @param clientSecret       the client secret to use when refreshing the access token.
+     * @param privateKey         the private key corresponding to the public key configured with Box Developer Edition.
+     * @param privateKeyPassword the password for the private key.
+     * @return a new instance of BoxAPIConnection.
+     * @throws JoseException
+     * @throws IOException
+     */
+    public static BoxAPIConnection getAppUserConnection(String userId, String clientId, String clientSecret,
+        String privateKey, String privateKeyPassword) throws JoseException, IOException {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: Same as above. Update exceptions.

File: src/main/java/com/box/sdk/BoxAPIConnection.java [hunk 5]
```diff
@@ -584,4 +691,41 @@ String lockAccessToken() {
     void unlockAccessToken() {
         this.refreshLock.readLock().unlock();
     }
+
+    String jwtConstructAssertion(String entityId, String entityType, String privateKey, String privateKeyPassword)
+        throws JoseException, IOException {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: Update exceptions.

**Comment by gcurtis** [comment]
Thanks! I left some inline comments on versioning and the exceptions that are being thrown.

I noticed that `BoxAPIConnection` doesn't support refresh, which means there isn't an easy way of staying authenticated if the access token expires (which I believe happens after an hour). Instead, the developer needs to keep calling `authenticate()` and providing their keys again. I think it might be better to support refresh by just requesting a new access token automatically using the original assertion.

What are your thoughts?

**Comment by gcurtis** [comment]
I made a change to `build.gradle` to address the checkstyle errors, so rebasing off of master should fix it. It looks like there still some legitimate warnings though.

**Comment by bgoldman** [comment]
1. Your change requests have been implemented.
2. I moved the code for Box Developer Edition to its own class, because we needed to store the variables for refreshing.

**Comment by gcurtis** [comment]
These changes have been merged in manually and are now on the master branch.
